feature_engine/prediction/target_mean_prediction.py

```diff
@@ -0,0 +1,212 @@
+    """
+
+    Parameters
+    ----------
+    variables: list, default=None
+        The list of variables to evaluate. If None, the transformer will evaluate all
+        variables in the dataset.
+
+    bins: int, default=5
+        If the dataset contains numerical variables, the number of bins into which
+        the values will be sorted.
+
+    strategy: str, default='equal_width'
+        Whether to create the bins for discretization of numerical variables using
+        equal width ('equal_width') or equal frequency ('equal_frequency').
+
+    ignore_format: bool, default=False
+        Whether the format in which the categorical variables are cast should be
+        ignored. If False, the encoder will automatically select variables of type
+        object or categorical, or check that the variables entered by the user are of
+        type object or categorical. If True, the encoder will select all variables or
+        accept all variables entered by the user, including those cast as numeric.
+
+    Attributes
+    ----------
+
+
+    Methods
+    -------
+    fit:
+
+    predict:
+
+    Notes
+    -----
+
+
+    See Also
+    --------
+    feature_engine.encoding.MeanEncoder
+    feature_engine.discretisation.EqualWidthDiscretiser
+    feature_engine.discretisation.EqualFrequencyDiscretiser
+
+    References
+    ----------
+
+
+    """
+
+    def __init__(
+        self,
+        variables: Union[None, int, str, List[Union[str, int]]] = None,
+        bins: int = 5,
+        strategy: str = "equal-width",
+        ignore_format: bool = False,
+    ):
+
+        if not isinstance(bins, int):
+            raise TypeError("'bins' only accepts integers.")
+
+        if strategy not in ("equal-width", "equal-distance"):
+            raise ValueError(
+                "strategy must be 'equal-width' or 'equal-distance'."
+            )
+
+        if not isinstance(ignore_format, bool):
+            raise ValueError("ignore_format takes only booleans True and False")
+
+        self.variables = _check_input_parameter_variables(variables)
+        self.bins = bins
+        self.strategy = strategy
+        self.ignore_format = ignore_format
+
+    def fit(self, X: pd.DataFrame, y: pd.Series = None) -> pd.DataFrame:
+        """
+        Fit predictor per variables.
+
+        QUESTION: Does X accept the entire dataframe or just on feature?
+
+        Parameters
+        ----------
+        X : pandas dataframe of shape = [n_samples, n_features]
+            The training input samples.
+
+        y : pandas series of shape = [n_samples,]
+            The target variable.
+        """
+        # check if dataframe
+        _is_dataframe(X)
+
+        if not isinstance(y, pd.Series):
+            y = pd.Series(y)
+
+        # check variables
+        self.variables_ = _find_all_variables(X, self.variables)
+
+        # check if df contains na
+        _check_contains_na(X, self.variables_)
+
+        self.n_features_in_ = X.shape[1]
```

Comment:
I moved the code to the end. Two questions:

- Why do we not check for NaN values prior to passing `X` to the encoder and/or discretisers?
- Does `self.n_features_in_ = X.shape[1]` go at the end of `fit()` because the encoder and/or discretisers will transform `X` into `X_new` and the transformation may change the dataframe's shape?

Reply:
On the check of nan point, if I remember correctly, we did not include it, because the transformer you are using in the pipeline already do those checks. So it is redundant.

On the second point, the `check_is_fitted()` function from sklearn that we use in the transform method looks for trailing underscores to decide if the transformer was fit or not. If we do the `self.n_features_in_` at the beginning, this attribute will have a value, even if the functionality below fails, so the check_is_fitted() will think the transformer was indeed fit, when actually, it raised an error.

Putting that check at the end, means that everything else has worked, should the check_is_fitted() found trailing underscores.